### DISCUSSION TITLE
Some initial ideas

### DISCUSSION THREAD

**Discussion by rediguana** [tphakala/birdnet-go#352] [General]
Firstly, only found this yesterday, and been amazed at how easy it was to get setup. This is a great project, and exactly the solution I was looking for. It reeks of potential and I'm quite excited by what the developers have achieved already :)

I started with the docker run command with rtsp to a couple of the security cameras and it installed the dev branch flawlessly. With a little playing, I was then able to setup a more persistent container and volumes using Portainer pretty easily. Happy to document how to setup in Portainer if anyone would find that useful.

Some of the initial things that I could see that may make for a better onboarding/setup process and initial use could include...

1. **Start container with no sound sources**. I tried to start the container with no RTSP streams, hoping I could then manually add them from Settings > Audio Capture, but the logging prevented the successful start of the container. The onboarding might be easier for those with cameras/RTSP streams available to start with no sound source, and then add the streams in the UI. The only way to start the container was with the commandline realtime rtsp tag with at least one stream. I don't know whether this is a docker thing that has to be called from the commandline?
2. **Name RTSP streams in Audio Settings**. It would be useful, when configuring RTSP streams to add a name to distinguish the streams. Right now I've got three streams and it would be great to have them named, which leads to the below.
3. **Add source to Recent Detections**. I don't know if this is possible, but if it is, it would be great to know which source a detection came from. I can then tell the approximate area that the detection came from.
4. **Whitelist/Blacklist detections**. As someone from New Zealand, there are some pretty obvious false positives coming up in the analysis. Is there any way that we can blacklist birds that we know we don't get. It would be great to also whitelist a lot of the birds that are common in our area. I did see in my surfing yesterday and the day before, that I think the source project is looking for more NZ bird recordings for training.

The ease with which I was able to leverage existing IT at home to implement BirdNET-Go was awesome. You've done some awesome work getting to this point, and it looks like it has a lot of potential! I am stoked that you have made the effort to get this working in docker - as I was not keen to either get more hardware (RPi) or tack BirdNET onto my Home Assistant instance.

Ngā mihi nui

Gavin

**Comment by tphakala** [comment]
"Name RTSP streams in Audio Settings" this is on todo list, "Add source to Recent Detections" and this is related to same feature.

"Whitelist/Blacklist detections." this is already possbile, it just lacks user interface elements, for now you can apply species specific settings through species_config.csv file which should be placed next to config.yaml, https://github.com/tphakala/birdnet-go/blob/main/doc/species_config.csv. BirdNET-Go restart is required to reload settings in this file.

"Start container with no sound sources" this should already be possible, did you install BirdNET-Go using install.sh?

**Reply by isZumpo** [reply]
> "Start container with no sound sources" this should already be possible, did you install BirdNET-Go using install.sh?

The issue appears to be that when starting the docker container without sending in the --device /dev/snd or having no mics connected, birdnet-go will shut itself down like this:

```
Available Capture Sources:
  0: Discard all samples (playback) or generate zero samples (capture), null
Error selecting capture source: no audio devices available in container
Please map host audio devices by running docker with: --device /dev/snd
Instructions for running BirdNET-Go in Docker are at https://github.com/tphakala/birdnet-go/blob/main/doc/installation.md
```

The snippet producing this error is in capture.go:

```go
// Check if running in container and only null device is available
if conf.RunningInContainer() && len(infos) == 1 && strings.Contains(infos[0].Name(), "Discard all samples") {
  return captureSource{}, fmt.Errorf(
	  "no audio devices available in container\n" +
		  "Please map host audio devices by running docker with: --device /dev/snd\n" +
		  "Instructions for running BirdNET-Go in Docker are at https://github.com/tphakala/birdnet-go/blob/main/doc/installation.md")
}
```

I tried to build a container image without this if  condition producing errors like this:


```go
if conf.RunningInContainer() && len(infos) == 1 && strings.Contains(infos[0].Name(), "Discard all samples") {
  fmt.Printf(
	  "no audio devices available in container\n" +
		  "Please map host audio devices by running docker with: --device /dev/snd\n" +
		  "Instructions for running BirdNET-Go in Docker are at https://github.com/tphakala/birdnet-go/blob/main/doc/installation.md\n")
  return captureSource{}, nil
}
```

then we end with an internal error instead....


```bash
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM default
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_card_id returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM dsnoop
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_card_id returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM dsnoop:0
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_card_id returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5180:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5703:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2666:(snd_pcm_open_noupdate) Unknown PCM dsnoop:0,0
ALSA lib confmisc.c:165:(snd_config_get_card) Cannot get card index for 0
ALSA lib confmisc.c:165:(snd_config_get_card) Cannot get card index for 0
ALSA lib confmisc.c:165:(snd_config_get_card) Cannot get card index for 0
```

Perhaps a method of selecting a nil audio source could be added to the codebase? Where if a nil audio source was selected it will try to not perform any audio capture until a audio source is configured by the user?

**Reply by rediguana** [reply]
> "Start container with no sound sources" this should already be possible, did you install BirdNET-Go using install.sh?

No, I just pulled the dev docker container from GitHub.

**Reply by jkrauska** [reply]
the install.sh has some nice optimizations for running in docker and tuning to match your cpu..  highly reccomend.

**Reply by tphakala** [reply]
@rediguana 
`Whitelist/Blacklist detections. `

Species settings UI is now included in latest code base, it contains include and exclude specis lists. 

- Include species list allows to override range filter (BirdNET model for propable species in region) by adding species to include for detection
- Exclude species list allows to block species from detection

**Reply by tjcuddihy** [reply]
> "Start container with no sound sources" this should already be possible, did you install BirdNET-Go using install.sh?

I also have encountered this challenge. My use-case is installing on [Unraid](https://unraid.net/) as a processor for RTSP stream from a RPi Zero 2w.

In Unraid, the expected workflow for docker images is to use a GUI page to define the `docker run ...` command that is used to manage the container.
<img width="948" alt="image" src="https://github.com/user-attachments/assets/a7306edb-daed-471c-a957-bffd7b198b46" />

Using the `instal.sh` doesn't quite work within this system (and I image it would be the same for any NAS users). Being able to `docker pull ...` and have a useable BirdNet-Go container would be a great workflow.

**Reply by tphakala** [reply]
I will fix this and release update during this weekend

**Reply by tphakala** [reply]
Any restrictions to start without configured audio sources are now removed, also any new RTSP streams are started and removed are stopped at runtime without requirement to restart BirdNET-Go. These changes are currently in nightly builds (including container images), once I am confident that nothing major didn't break I'll release 0.6.1.

**Reply by xconverge** [reply]
@tphakala I am unable to run with only a RTSP stream with nightly-20250211, my config works fine with nightly-20250209

See my log here https://github.com/tphakala/birdnet-go/issues/415#issuecomment-2652253964

**Reply by tphakala** [reply]
okay yeah, I see that there is still issues left .. I'll fix these later today. However I don't understand how there could be regresssion related to this between those releases.

**Reply by xconverge** [reply]
> okay yeah, I see that there is still issues left .. I'll fix these later today. However I don't understand how there could be regresssion related to this between those releases.

You might be right unless there was some race condition or something more intermittent. I ended up needing to nuke my db and start with a clean slate back on 0.6.0. I had been running nightly-20250209 for 48 hours with no issues though, not really sure, but I agree with you to not necessarily use those builds to bisect the issue!
